### ISSUE TITLE
Provide mechanism to run GrumPHP without Git for a list of specified files

### ISSUE THREAD

**Issue by Bilge** [phpro/grumphp#394]
| Q               | A
| --------------- | ---
| Version         | n/a
| Bug?            | no
| New feature?    | yes
| Question?       | no
| Documentation?  | no
| Related tickets | n/a

GrumPHP currently relies on Git to feed it the list of files to process. However, it is a common requirement to want to reuse GrumPHP's functionality in contexts where the list of files may come from another source. It should be possible for the user (or a script) to specify a list of files to check on GrumPHP's command line.

An example use case might be compiling a list of all files changed between two commit refs and asking GrumPHP to validate them.

As a consideration towards future work, consider how support might be added to specify directories that are recursively searched in addition to specific files.

**Comment by veewee** [comment]
Sounds like a good but rather complex idea. Thanks for adding it to the feature list :)

**Comment by thomscode** [comment]
Theoretically it's not too complicated. If GrumPHP had a way to accept a list of files from the command line `grumphp run <space separated file/folder list>` then the burden of providing that listing would be on the user. If the user wants the list to be fed from a list of files in a diff, then that user could write a diff parser to pull the file listing.

Slightly more complicated is to allow plugins that could parse input and provide a list of files. In this way users can write plugins themselves that fit their needs. If you pursue this way, then the file list is simply another parser that explodes the space separated list into an array.

Anyway, just a couple of ideas.

**Comment by UrGuardian4ngel** [comment]
Or maybe you could read the file list from `STDIN` when a certain flag is defined, like e.g. `-i|--from-stdin`?

```bash
# from e.g. a static file list
$ cat file-list.txt | grumphp run --from-stdin

# extracted from a git diff
$ git diff HEAD~3... --name-only | grumphp run --from-stdin

# or generated by any other tool (in this case, silver searcher)
$ ag -l --php -- src/AppBundle/SomeComponent | grumphp run --from-stdin
```

**Comment by veewee** [comment]
@UrGuardian4ngel : in the pre-commit task, we allready support this. We do need a git diff at this moment though. (See `ChangedFiles` locator for more details)

**Comment by UrGuardian4ngel** [comment]
Oh yes. But I was assuming he was talking about the `grumphp run` command, as per the issue title:

> Provide mechanism to **run GrumPHP** without Git **for a list of specified files**

which is using `RegisteredFiles` (`$ git ls-files`), and depends on a `Git\Repository`.

– P.S.: Didn't see the `grumphp git:pre-commit` internals yet 

---

IMO this "non-git command" (`grumphp run`) would not necessarily require a _git repo_ to work. If you were e.g. manually able to pass a file list in (via pipe, variadic argument, ...). It could still default to `$ git ls-files` though, if no file list is otherwise present.
Basically applying the same trick as your `Git\PreCommitCommand` – with the difference that `grumphp run` would accept a file list, and not a raw `git diff` output (as `grumphp git:pre-commit` does).

Granted, all other functionalities in the _"git"_ command namespace (`grumphp git:*`) are obviously coupled to git functionalities, so there's also that... :)

**TL;DR:** Maybe we could apply a similar trick to the `grumphp run` command?

**Comment by paslandau** [comment]
Are there any plans to move forward with this request? I second that it would be very helpful to decouple `grumphp run` from the git diff. We're using a subset of the provided tools anyway and use them e.g. for code review. Having the ability to pass folders/files thus is a requirement for us and grumphp would let us have a one consistent place for qa configuration.

**Comment by hkirsman** [comment]
+1 for this!

**Comment by tarunjangra** [comment]
+1

**Comment by veewee** [comment]
This suggestion has been added through #801. Feel free to test it out.
